modules/infra/control/port.c

```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
 	.to_api = port_to_api,
 };
 
+int gr_port_mcast_addr_add(struct iface_info_port *p, struct rte_ether_addr *mcast_addr) {
+	struct rte_ether_addr mac[GR_PORT_NUM_MCAST_ADDR];
+	uint16_t count = 0;
+	bool found = false;
+	int ret;
+	int i;
+
+	if (p->flags & GR_PORT_F_ALLMULTI_TOO_MANY_MCAST) {
+		// allmulti is already enabled, internal mcast addr list overflow
+		return 0;
+	}
+
+	if (mcast_count(p) == GR_PORT_NUM_MCAST_ADDR) {
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+		goto enable_allmulti;
+	}
+
+	for (i = 0; i < GR_PORT_NUM_MCAST_ADDR && !found; i++) {
+		if (rte_is_same_ether_addr(mcast_addr, &p->multicast[i].mac)) {
+			LOG(DEBUG, "multicast address already filtered by port %d", p->port_id);
+			p->multicast[i].refcnt++;
+			return 0;
+		} else if (p->multicast[i].refcnt == 0) {
+			memcpy(&p->multicast[i].mac, mcast_addr, sizeof(struct rte_ether_addr));
+			p->multicast[i].refcnt = 1;
+			found = true;
+		}
+	}
+
+	if (i == GR_PORT_NUM_MCAST_ADDR) {
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+		LOG(INFO,
+		    "Too many multicast addresses requested for %s",
+		    port_get_iface(p->port_id)->name);
+		goto enable_allmulti;
+	}
+
+	count = mcast_count(p);
+	for (i = 0; i < count; i++) {
+		memcpy(&mac[i], &p->multicast[i].mac, sizeof(struct rte_ether_addr));
+	}
+
+	ret = rte_eth_dev_set_mc_addr_list(p->port_id, mac, count);
+
+	if (ret == 0) {
+		return 0;
+	} else if (ret == -ENOSPC) {
+		LOG(WARNING,
+		    "Enabling mc addr list: no space left on iface %s.",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+	} else if (ret == -ENOTSUP) {
+		LOG(WARNING,
+		    "Enabling mc addr list not supported for iface %s",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_DEVICE_LIMIT;
+	}
```

Comment:
Minor style preferences:

* Do not use capital letters nor punctuation in log messages.
* Prefer shorter messages if possible.
* Maybe a switch case would be better here.